### PR TITLE
Leah - Space

### PR DESCRIPTION
**Pull request by leahwho**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The model is the setup for the database - it controls how columns are named.
Describe in your own words what the Controller is doing in Rails | The controller is the conductor - it talks to the model and the view in order to send back a response to the client/browser/user.  It defines what actions to take when it receives a request from the router.
Describe in your own words what the View is doing in Rails | The view is in charge of what is displayed in the response.  It controls how things look and holds all the templates for each route.
Describe an edge-case controller test you wrote | I passed an invalid ID to test a redirection.
What is the purpose of using strong params? (i.e. the params method in the controller) | Strong params allow us to DRY up our code and not have to type so much.  If you create a params method, you can define what keys are allowable to access in form submission.
How are Rails migrations related to Rails models? | Migrations are how we can edit the schema/model of the database.  If you need to create a table, change the data type or column name, or remove a column, you create a migration which allows you to edit what you need to change.
Describe one area of Rails that are still unclear on | I need to get a little more solid on redirecting, building custom routes, and increasing the complexity of what the model is doing.

### CODE REVIEW

File: test/controllers/tasks_controller_test.rb [hunk 1]
```diff
@@ -72,53 +67,105 @@
...
+
+    it 'redirects if given invalid id' do
+      delete task_path(-10)
+
+      must_respond_with :redirect
+    end
   end
   
   # Complete for Wave 4
   describe "toggle_complete" do
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:149]: Nice work fleshing out these tests. Make sure to add an expectation for `must_redirect_to`

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    <%= csp_meta_tag %>
+
+    <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
+  </head>
+
+  <section class='app-container'>
+    <header>
+    <h1 class='to-do-title'>To Do List</h1>
+    <nav class='nav-links-box'>
```
Comment (beccaelenzil) [inline app/views/layouts/application.html.erb:15]: Great use of layouts.

File: app/views/tasks/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<h2><%= @task.name %></h2>
```
Comment (beccaelenzil) [inline app/views/tasks/show.html.erb:1]: User experience is great. Nice work on this view!

File: app/views/tasks/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+<h2>Enter Your New Task:</h2>
```
Comment (beccaelenzil) [inline app/views/tasks/new.html.erb:1]: Nice work using partial views. As we discussed in round tables, you could include the `<h2>` element in your `_form` view by adding more keys to the `locals` hash.

File: app/controllers/tasks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  
+  
+  def mark_complete
+    id = params[:id]
+    @task = Task.find_by(id: id)
+    
+    if @task.nil?
+      redirect_to tasks_path
+      return
+    else @task.completed_at = Time.now
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:87]: Minor style note: `@task.completed_at = Time.now` should be on a new line.

**Review by beccaelenzil — COMMENTED** [review]
Great work on your first rails project! It is clear that the learning goals for your creating your first rails app were met. Nice work using partial views, fleshing out the tests for changes to the db, and creating a stylin' app. You're off to a great start. Keep up the hard work!

**Comment by beccaelenzil** [comment]
# Task List




## Major Learning Goals/Code Review





| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| At least 6 commits with meaningful commit messages | ✔️
| Routes follow RESTful conventions | ✔️
| Uses named routes (like `_path`) | ✔️
| Creates Models and migrations | ✔️
| Creates styled views | ✔️
| Handles errors like nonexistant tasks | ✔️
| Uses `form_with` to render forms in Rails | ✔️

## Functional Requirements/Manual Testing

| Functional Requirement | yes/no |
| --- | --- |
| Successfully handles index & show | ✔️
| index & show tests pass | ✔️
| Successfully handles: New, Create | ✔️
| New, Create tests pass | ✔️
| Successfully handles: Edit, Update | ✔️
| Successfully handles: Destroy, Task Complete | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 5+ in Functional Requirements | ✔️






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
